pkg/queue/cluster_queue_impl.go

```diff
@@ -55,39 +72,99 @@ func (c *ClusterQueueImpl) Cohort() string {
+		// to potentially become admissible.
+		if equality.Semantic.DeepEqual(oldInfo.Obj.Spec, wInfo.Obj.Spec) {
+			c.inadmissibleWorkloads[key] = wInfo
+			return
+		}
+		// otherwise move or update in place in the queue.
+		delete(c.inadmissibleWorkloads, key)
+	}
+	c.heap.PushOrUpdate(wInfo)
+}
+
+func (c *ClusterQueueImpl) Delete(w *kueue.Workload) {
+	delete(c.inadmissibleWorkloads, workload.Key(w))
+	c.heap.Delete(workload.Key(w))
+}
+
 func (c *ClusterQueueImpl) DeleteFromQueue(q *Queue) {
+	for _, w := range q.items {
+		key := workload.Key(w.Obj)
+		if wl := c.inadmissibleWorkloads[key]; wl != nil {
+			delete(c.inadmissibleWorkloads, key)
+		}
+	}
 	for _, w := range q.items {
 		c.Delete(w.Obj)
 	}
 }
 
-// pushIfNotPresent pushes the workload to ClusterQueue.
-// If the workload is already present, returns false. Otherwise returns true.
-func (c *ClusterQueueImpl) pushIfNotPresent(info *workload.Info) bool {
-	return c.heap.PushIfNotPresent(info)
-}
+// RequeueIfNotPresent inserts a workload that cannot be admitted into
+// ClusterQueue, unless it is already in the queue. If immediate is true,
+// the workload will be pushed back to heap directly. If not,
+// the workload will be put into the inadmissibleWorkloads.
+func (c *ClusterQueueImpl) RequeueIfNotPresent(wInfo *workload.Info, immediate bool) bool {
+	key := workload.Key(wInfo.Obj)
+	if immediate {
+		// If the workload was inadmissible, move it back into the queue.
+		inadmissibleWl := c.inadmissibleWorkloads[key]
+		if inadmissibleWl != nil {
+			wInfo = inadmissibleWl
+			delete(c.inadmissibleWorkloads, key)
+		}
+		return c.heap.PushIfNotPresent(wInfo)
+	}
 
-func (c *ClusterQueueImpl) PushOrUpdate(info *workload.Info) {
-	c.heap.PushOrUpdate(info)
-}
+	if c.inadmissibleWorkloads[key] != nil {
+		return false
+	}
 
-func (c *ClusterQueueImpl) Delete(w *kueue.Workload) {
-	c.heap.Delete(workload.Key(w))
+	if data := c.heap.GetByKey(key); data != nil {
+		return false
+	}
+
+	c.inadmissibleWorkloads[key] = wInfo
+
+	return true
 }
 
-func (c *ClusterQueueImpl) RequeueIfNotPresent(wInfo *workload.Info, _ bool) bool {
-	return c.pushIfNotPresent(wInfo)
+// QueueInadmissibleWorkloads moves all workloads from inadmissibleWorkloads to heap.
+// If at least one workload is moved, returns true. Otherwise returns false.
+func (c *ClusterQueueImpl) QueueInadmissibleWorkloads(client client.Client) bool {
+	if len(c.inadmissibleWorkloads) == 0 {
+		return false
+	}
+
+	inadmissibleWorkloads := make(map[string]*workload.Info)
+	for key, wInfo := range c.inadmissibleWorkloads {
+		ns := corev1.Namespace{}
+		err := client.Get(context.Background(), types.NamespacedName{Name: wInfo.Obj.Namespace}, &ns)
+		if err != nil || !c.namespaceSelector.Matches(labels.Set(ns.Labels)) {
+			inadmissibleWorkloads[key] = wInfo
+		} else {
+			c.heap.PushIfNotPresent(wInfo)
+		}
+	}
+
+	c.inadmissibleWorkloads = inadmissibleWorkloads
+	return true
```

Comment:
This might be false if all workloads not match the namespaceSelector or `PushIfNotPresent` always returns false.

Reply:
true, updated.